Partially implement #74

jroelofs opened the pull request:
Securely erase keys from memory when they go out of scope. Might have a
performance impact, which I haven't measured (perf measurements aren't
generally reliable on laptops).

I wasn't entirely happy with having to add the `scrubbed::inner()` methods, but some `static_assert`s in the serialization stuff complains that the wrapper class's dtor causes there to be padding, for reasons I don't yet understand. I considered hiding it slightly with an `operator*()`, but wasn't sure what the norms were in this project toward that. I'm happy to take advice/feedback here.

jroelofs commented:
Also, I don't have a windows machine to test this on, so that would be greatly appreciated.

stoffu commented:
How about pretending that `secret_key` is a POD type by casting it into `ec_scalar`? This eliminates the need for the `inner()` function (while slightly lying about the POD-ness):
```diff
diff --git a/src/crypto/crypto.cpp b/src/crypto/crypto.cpp
index 95ba3482..db4f5bac 100644
--- a/src/crypto/crypto.cpp
+++ b/src/crypto/crypto.cpp
@@ -43,6 +43,13 @@
 #include "crypto.h"
 #include "hash.h"
 
+#ifdef WIN32
+    // For SecureZeroMemory
+    #include <windows.h>
+#else
+    #include <strings.h>
+#endif
+
 namespace crypto {
 
   using std::abort;
@@ -59,6 +66,20 @@ namespace crypto {
 
   boost::mutex random_lock;
 
+#ifndef WIN32
+  static void *(*const volatile memset_vp)(void *, int, size_t) = memset;
+#endif
+
+  void scrub_memory(void *data, size_t len) {
+#ifdef WIN32
+    SecureZeroMemory(data, len);
+#else
+    // On platforms without a memset_s/explicit_bzero/SecureZeroMemory, call
+    // through a volatile ptr, which will not be optimized out by the compiler.
+    (*memset_vp)(data, 0, len);
+#endif
+  }
+
   static inline unsigned char *operator &(ec_point &point) {
     return &reinterpret_cast<unsigned char &>(point);
   }
diff --git a/src/crypto/crypto.h b/src/crypto/crypto.h
index abdea016..7c4e6b1e 100644
--- a/src/crypto/crypto.h
+++ b/src/crypto/crypto.h
@@ -36,9 +36,11 @@
 #include <boost/thread/lock_guard.hpp>
 #include <boost/utility/value_init.hpp>
 #include <boost/optional.hpp>
+#include <type_traits>
 #include <vector>
 
 #include "common/pod-class.h"
+#include "common/util.h"
 #include "generic-ops.h"
 #include "hex.h"
 #include "span.h"
@@ -52,6 +54,29 @@ namespace crypto {
 
   extern boost::mutex random_lock;
 
+  /// Destroy the contents of the pointed-to memory.
+  void scrub_memory(void *data, size_t len);
+
+  /// Scrubs data in the contained type upon destruction.
+  ///
+  /// Primarily useful for making sure that private keys don't stick around in
+  /// memory after the objects that held them have gone out of scope.
+  template <class T>
+  struct scrubbed : public T {
+    ~scrubbed() {
+      scrub();
+    }
+
+    /// Destroy the contents of the contained type.
+    void scrub() {
+      static_assert(std::is_pod<T>::value,
+                    "T cannot be auto-scrubbed. T must be POD.");
+      static_assert(std::is_trivially_destructible<T>::value,
+                    "T cannot be auto-scrubbed. T must be trivially destructable.");
+      scrub_memory(this, sizeof(T));
+    }
+  };
+
 #pragma pack(push, 1)
   POD_CLASS ec_point {
     char data[32];
@@ -65,9 +90,7 @@ namespace crypto {
     friend class crypto_ops;
   };
 
-  POD_CLASS secret_key: ec_scalar {
-    friend class crypto_ops;
-  };
+  using secret_key = scrubbed<ec_scalar>;
 
   POD_CLASS public_keyV {
     std::vector<public_key> keys;
@@ -261,7 +284,38 @@ namespace crypto {
     const signature *sig) {
     return check_ring_signature(prefix_hash, image, pubs.data(), pubs.size(), sig);
   }
+}
+
+// template specialization for crypto::secret_key so that it can pretend to be a POD type
+namespace epee
+{
+  namespace string_tools
+  {
+    template<class t_pod_type>
+    std::string pod_to_hex(const t_pod_type& s);
+    template<>
+    inline std::string pod_to_hex<crypto::secret_key>(const crypto::secret_key& s)
+    {
+      return pod_to_hex(static_cast<const crypto::ec_scalar &>(s));
+    }
+    //----------------------------------------------------------------------------
+    template<class t_pod_type>
+    bool hex_to_pod(const std::string& hex_str, t_pod_type& s);
+    template<>
+    inline bool hex_to_pod<crypto::secret_key>(const std::string& hex_str, crypto::secret_key& s)
+    {
+      return hex_to_pod(hex_str, static_cast<crypto::ec_scalar &>(s));
+    }
+  }
+  //----------------------------------------------------------------------------
+  template<>
+  span<const std::uint8_t> as_byte_span<crypto::secret_key>(const crypto::secret_key& src) noexcept
+  {
+    return as_byte_span(static_cast<const crypto::ec_scalar &>(src));
+  }
+}
 
+namespace crypto {
   inline std::ostream &operator <<(std::ostream &o, const crypto::public_key &v) {
     epee::to_hex::formatted(o, epee::as_byte_span(v)); return o;
   }
```

jroelofs commented:
I like it. It’s not lying any more than `inner()` was, and the
specialization avoids changing all of the places where it’s used.

On Thu, Nov 23, 2017 at 8:21 PM stoffu <notifications@github.com> wrote:

> How about pretending that secret_key is a POD type by casting it into
> ec_scalar? This eliminates the need for the inner() function (while
> slightly lying about the POD-ness):
>
> diff --git a/src/crypto/crypto.cpp b/src/crypto/crypto.cpp
> index 95ba3482..db4f5bac 100644--- a/src/crypto/crypto.cpp+++ b/src/crypto/crypto.cpp@@ -43,6 +43,13 @@
>  #include "crypto.h"
>  #include "hash.h"
>  +#ifdef WIN32+    // For SecureZeroMemory+    #include <windows.h>+#else+    #include <strings.h>+#endif+
>  namespace crypto {
>
>    using std::abort;@@ -59,6 +66,20 @@ namespace crypto {
>
>    boost::mutex random_lock;
>  +#ifndef WIN32+  static void *(*const volatile memset_vp)(void *, int, size_t) = memset;+#endif++  void scrub_memory(void *data, size_t len) {+#ifdef WIN32+    SecureZeroMemory(data, len);+#else+    // On platforms without a memset_s/explicit_bzero/SecureZeroMemory, call+    // through a volatile ptr, which will not be optimized out by the compiler.+    (*memset_vp)(data, 0, len);+#endif+  }+
>    static inline unsigned char *operator &(ec_point &point) {
>      return &reinterpret_cast<unsigned char &>(point);
>    }diff --git a/src/crypto/crypto.h b/src/crypto/crypto.h
> index abdea016..7c4e6b1e 100644--- a/src/crypto/crypto.h+++ b/src/crypto/crypto.h@@ -36,9 +36,11 @@
>  #include <boost/thread/lock_guard.hpp>
>  #include <boost/utility/value_init.hpp>
>  #include <boost/optional.hpp>+#include <type_traits>
>  #include <vector>
>
>  #include "common/pod-class.h"+#include "common/util.h"
>  #include "generic-ops.h"
>  #include "hex.h"
>  #include "span.h"@@ -52,6 +54,29 @@ namespace crypto {
>
>    extern boost::mutex random_lock;
>  +  /// Destroy the contents of the pointed-to memory.+  void scrub_memory(void *data, size_t len);++  /// Scrubs data in the contained type upon destruction.+  ///+  /// Primarily useful for making sure that private keys don't stick around in+  /// memory after the objects that held them have gone out of scope.+  template <class T>+  struct scrubbed : public T {+    ~scrubbed() {+      scrub();+    }++    /// Destroy the contents of the contained type.+    void scrub() {+      static_assert(std::is_pod<T>::value,+                    "T cannot be auto-scrubbed. T must be POD.");+      static_assert(std::is_trivially_destructible<T>::value,+                    "T cannot be auto-scrubbed. T must be trivially destructable.");+      scrub_memory(this, sizeof(T));+    }+  };+
>  #pragma pack(push, 1)
>    POD_CLASS ec_point {
>      char data[32];@@ -65,9 +90,7 @@ namespace crypto {
>      friend class crypto_ops;
>    };
>  -  POD_CLASS secret_key: ec_scalar {-    friend class crypto_ops;-  };+  using secret_key = scrubbed<ec_scalar>;
>
>    POD_CLASS public_keyV {
>      std::vector<public_key> keys;@@ -261,7 +284,38 @@ namespace crypto {
>      const signature *sig) {
>      return check_ring_signature(prefix_hash, image, pubs.data(), pubs.size(), sig);
>    }+}++// template specialization for crypto::secret_key so that it can pretend to be a POD type+namespace epee+{+  namespace string_tools+  {+    template<class t_pod_type>+    std::string pod_to_hex(const t_pod_type& s);+    template<>+    inline std::string pod_to_hex<crypto::secret_key>(const crypto::secret_key& s)+    {+      return pod_to_hex(static_cast<const crypto::ec_scalar &>(s));+    }+    //----------------------------------------------------------------------------+    template<class t_pod_type>+    bool hex_to_pod(const std::string& hex_str, t_pod_type& s);+    template<>+    inline bool hex_to_pod<crypto::secret_key>(const std::string& hex_str, crypto::secret_key& s)+    {+      return hex_to_pod(hex_str, static_cast<crypto::ec_scalar &>(s));+    }+  }+  //----------------------------------------------------------------------------+  template<>+  span<const std::uint8_t> as_byte_span<crypto::secret_key>(const crypto::secret_key& src) noexcept+  {+    return as_byte_span(static_cast<const crypto::ec_scalar &>(src));+  }+}
>  +namespace crypto {
>    inline std::ostream &operator <<(std::ostream &o, const crypto::public_key &v) {
>      epee::to_hex::formatted(o, epee::as_byte_span(v)); return o;
>    }
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/monero-project/monero/pull/2857#issuecomment-346734842>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADfnSX03rJdFEbAalcI-BiEW8HiKTK0Aks5s5jYdgaJpZM4QpT0J>
> .
>

moneromooo-monero commented:
I had a memwipe function waiting, now PR'd in https://github.com/monero-project/monero/pull/2860, which could be of use for this patch.

jroelofs commented:
@moneromooo-monero thanks! I'll rebase on top of that patch. Makes sense to share the `memwipe()` infra.

stoffu commented:
How about this alternative which is more explicit about our intention (to make scrubbed type look like POD) and makes the change even smaller:

```diff
--- a/src/crypto/crypto.h
+++ b/src/crypto/crypto.h
@@ -262,6 +285,19 @@ namespace crypto {
     return check_ring_signature(prefix_hash, image, pubs.data(), pubs.size(), sig);
   }
 
+}
+
+// Overlaods for crypto::scrubbed<T> so that it can pretend to be the contained
+// POD type in those contexts.
+namespace std
+{
+  template<class t_scrubbee>
+  struct is_pod<crypto::scrubbed<t_scrubbee>> {
+    static const bool value = is_pod<t_scrubbee>::value;
+  };
+}
+
+namespace crypto {
   inline std::ostream &operator <<(std::ostream &o, const crypto::public_key &v) {
     epee::to_hex::formatted(o, epee::as_byte_span(v)); return o;
   }
--- a/contrib/epee/include/span.h
+++ b/contrib/epee/include/span.h
@@ -108,7 +108,7 @@ namespace epee
   template<typename T>
   constexpr bool has_padding() noexcept
   {
-    return !std::is_pod<T>() || alignof(T) != 1;
+    return !std::is_pod<T>::value || alignof(T) != 1;
   }
 
   //! \return Cast data from `src` as `span<const std::uint8_t>`.
```

jroelofs commented:
We're allowed to add specializations in `namespace std`, but not new overloads, unfortunately. See: 17.6.2.4.1 [namespace.std]


How about something like (rough handwaving, haven't tried it):

```
namespace crypto {
  template<class t_scrubbee>
  struct is_pod<crypto::scrubbed<t_scrubbee>> {
    static const bool value = std::is_pod<t_scrubbee>::value;
  };
}
```

```
namespace epee {
   template<typename T>
   constexpr bool has_padding() noexcept {
     using std::is_pod<T>;
     using crypto::is_pod<T>;
     return !is_pod<T>::value || alignof(T) != 1;
   }
}
```

stoffu commented:
@jroelofs 
Thanks for your input. BTW you mistyped the section number of [namespace.std] as 17.6.2.4.1 which doesn't exist; here's an excerpt:

>  ### 17.6.4.2.1 Namespace `std` [namespace.std]
> The behavior of a C++ program is undefined if it adds declarations or definitions to namespace `std` or to a namespace within namespace `std` unless otherwise specified. A program may add a template specialization for any standard library template to namespace `std` only if the declaration depends on a user-defined type and the specialization meets the standard library requirements for the original template and is not explicitly prohibited.
> 
> The behavior of a C++ program is undefined if it declares
> * an explicit specialization of any member function of a standard library class template, or
> * an explicit specialization of any member function template of a standard library class or class template, or
> * an explicit or partial specialization of any member class template of a standard library class or class template.
> 
> A program may explicitly instantiate a template defined in the standard library only if the declaration depends on the name of a user-defined type and the instantiation meets the standard library requirements for the original template.
> 
> A translation unit shall not declare namespace `std` to be an inline namespace (7.3.1).

http://www.open-std.org/JTC1/SC22/WG21/docs/papers/2013/n3797.pdf

Just like how we define specializations of class template `std::hash<T>` for using unordered containers with custom types, the above code just defines specialization of class template `std::is_pod<T>` for our custom type `crypto::scrubbed<T>`. I don't see it meets any of the UB conditions (i.e. `std::pod<T>::value` is just a member data, not a member function nor a member class).

The code snippet you suggested gave me this compile error:
``` 
error: using declaration cannot refer to a template specialization
    using std::is_pod<T>;
               ^     ~~~
```

jroelofs commented:
Oh, right. Brain fart... it's not an overload. Not sure why I thought it was a function.

jroelofs commented:
I'll incorporate that into this patch, probably this evening or tomorrow.

jroelofs commented:
@stoffu I'm not completely familiar with the local norms w.r.t. approved commits yet... is there anything else I need to do to get this merged? (I suppose I need the PR it's based on to be accepted first)

stoffu commented:
@jroelofs 
Your patch seems ready for merging in my view, but at least @moneromooo-monero needs to confirm it as well. And yes, it makes sense to wait for the merge of the memwipe PR #2860 on which this patch is based.

Also, there are quite a few commits in this patch, which is fine if they're well justified. Otherwise, if you think some commits can be squashed, that is appreciated (and this is usually what @moneromooo-monero asks for most PRs in general).

Also it's important to keep the patch always mergeable to the latest master (by rebasing) every time @fluffypony merges other PRs. If merge conflicts are found, he'll ask you anyway to rebase, though.

jroelofs commented:
Perfect. That matches my normal workflow. Thanks for the review!

moneromooo-monero commented:
Agreed about squashing things. I like to have a patch per self contained "thing" in general. Makes it easier to use git history later. Patches usually stay up for revewing for a while before being being merged, that doesn't necessarily mean you have anything more to do. When you rebase on top of the memwipe patch, pleae make sure you don't have a merge commit left.

jroelofs commented:
Squashed, rebased, and added a couple more uses of `tools::scrubbed`. Also added `scrubbed_arr<T, N>` since the pattern of doing this on arrays is pretty common.

jroelofs commented:
Rebased again.
